include/dgl/runtime/ndarray.h

```diff
@@ -415,6 +415,9 @@ struct NDArray::Container {
   std::vector<int64_t> stride_;
   /*! \brief The internal array object */
   std::atomic<int> ref_counter_{0};
+
+  bool pinned_by_dgl_{false};
+  bool from_tensor_dispatcher_{false};
```

Comment:
Why need this flag?

Reply:
This is to avoid unpinning temporary NDArrays in such a case:
```python
def pin_memory_inplace(tensor):
    F.zerocopy_to_dgl_ndarray_for_write(tensor).pin_memory_()
```

1. However, with it, PyTorch tensors will not be automatically unpinned, which means
```python
tensor = torch.ones(10, 10)
nd = F.zerocopy_to_dgl_ndarray_for_write(t)
nd.pin_memory_()

del tensor
del nd  # will not be unpinned
```
we have to use
```python
nd = ndarray.empty(10, 10)
tensor = F.zerocopy_from_dgl_ndarray(nd)
nd.pin_memory_()

del nd  # will be unpinned
del tensor
```

2. If I remove it, we have to change `pin_memory_inplace()` to
```python
def pin_memory_inplace(tensor):
    nd = F.zerocopy_to_dgl_ndarray_for_write(tensor)
    nd.pin_memory_()
    return nd
```
and require people to save `nd` in somewhere.

I prefer the latter now. How do you think about it? @BarclayII

Reply:
Yeah it seems that 2 is better since users are unlikely to call `pin_memory_inplace` by themselves.

Reply:
Updated.

Reply:
Could you fix the DataLoader's use_uva option to store the NDArray handle as well?

Reply:
I see.  Makes sense.  Note that adding new attributes in Columns without will cause #4107 however unless we override `__getstate__` and `__setstate__`.

Reply:
I see.  Makes sense.  Note that adding new attributes in Columns without will cause #4107 however unless we override `__getstate__` and `__setstate__`.